// webdriver-image-comparison/build/clientSideScripts/getAndroidStatusAddressToolBarOffsets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getAndroidStatusAddressToolBarOffsets(androidOffsets, _a) {
    var isHybridApp = _a.isHybridApp, isLandscape = _a.isLandscape;
    var _b = window.screen, height = _b.height, width = _b.width;
    var innerHeight = window.innerHeight;
    var match = navigator.appVersion.match(/Android (\d+).?(\d+)?.?(\d+)?/);
    var majorVersion = parseInt(match[1], 10);
    var versionOffsets = androidOffsets[majorVersion];
    var deviceHeight = isLandscape && width > height ? width : height;
    var deviceWidth = isLandscape && height > width ? height : width;
    var statusAddressBarHeight = versionOffsets.STATUS_BAR + (isHybridApp ? 0 : versionOffsets.ADDRESS_BAR);
    var toolBarHeight = height - innerHeight - statusAddressBarHeight;
    if (toolBarHeight < 0) {
        toolBarHeight = versionOffsets.TOOL_BAR;
    }
    return {
        safeArea: 0,
        screenHeight: deviceHeight,
        screenWidth: deviceWidth,
        statusAddressBar: {
            height: statusAddressBarHeight,
            width: width,
            x: 0,
            y: 0,
        },
        sideBar: { height: 0, width: 0, x: 0, y: 0 },
        toolBar: {
            height: toolBarHeight,
            width: width,
            x: 0,
            y: height - toolBarHeight,
        },
    };
}
exports.default = getAndroidStatusAddressToolBarOffsets;
