// webdriver-image-comparison/build/clientSideScripts/getIosStatusAddressToolBarOffsets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getIosStatusAddressToolBarOffsets(iosOffsets, isLandscape) {
    var _a = window.screen, width = _a.width, height = _a.height;
    var innerHeight = window.innerHeight;
    var isIphone = width < 1024 && height < 1024;
    var deviceType = isIphone ? 'IPHONE' : 'IPAD';
    var orientationType = isLandscape ? 'LANDSCAPE' : 'PORTRAIT';
    var defaultPortraitHeight = isIphone ? 667 : 1024;
    var portraitHeight = width > height ? width : height;
    var deviceHeight = isLandscape ? width : height;
    var deviceWidth = isLandscape ? height : width;
    var offsetPortraitHeight = Object.keys(iosOffsets[deviceType]).indexOf(portraitHeight.toString()) > -1 ? portraitHeight : defaultPortraitHeight;
    var currentOffsets = iosOffsets[deviceType][offsetPortraitHeight][orientationType];
    var osVersion = parseInt(navigator.appVersion.match(/(?:OS |Version\/)(\d+)(?:_|\.)(\d+)(?:_|\.)?(\d+)?/)[1], 10);
    var statusBarHeight = currentOffsets.STATUS_BAR;
    var isIpadPro129FirstGeneration = false;
    var safeArea = currentOffsets.SAFE_AREA;
    if ((deviceHeight === 1366 || deviceWidth === 1366) &&
        deviceType === 'IPAD' &&
        innerHeight + currentOffsets.ADDRESS_BAR + currentOffsets.STATUS_BAR > deviceHeight) {
        statusBarHeight = 20;
        isIpadPro129FirstGeneration = true;
    }
    if ((deviceHeight === 896 || deviceWidth === 896) && deviceType === 'IPHONE' && osVersion === 13) {
        if (isLandscape) {
            safeArea = 44;
        }
        else {
            statusBarHeight = 44;
        }
    }
    var addressBarOnTop = (!isLandscape && isIphone && osVersion < 15) || isLandscape || !isIphone;
    var statusAddressBarHeight = statusBarHeight + (addressBarOnTop ? currentOffsets.ADDRESS_BAR : 0);
    var toolBarHeight = height - innerHeight - statusAddressBarHeight;
    var toolBarOffsets = isLandscape || toolBarHeight <= 0
        ?
            isIpadPro129FirstGeneration
                ? { height: 0, width: 0, x: 0, y: 0 }
                : currentOffsets.HOME_BAR
        : {
            height: toolBarHeight,
            width: deviceWidth,
            x: 0,
            y: height - toolBarHeight,
        };
    var sideBarOffsets = isLandscape && !isIphone
        ? {
            height: deviceHeight - statusAddressBarHeight,
            width: deviceWidth - document.documentElement.scrollWidth,
            x: 0,
            y: statusAddressBarHeight,
        }
        : { height: 0, width: 0, x: 0, y: 0 };
    return {
        safeArea: safeArea,
        screenHeight: deviceHeight,
        screenWidth: deviceWidth,
        sideBar: sideBarOffsets,
        statusAddressBar: {
            height: statusAddressBarHeight,
            width: deviceWidth,
            x: 0,
            y: 0,
        },
        toolBar: toolBarOffsets,
    };
}
exports.default = getIosStatusAddressToolBarOffsets;
