// webdriver-image-comparison/build/clientSideScripts/getScreenDimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getScreenDimensions() {
    var body = document.body;
    var html = document.documentElement;
    var bodyDimensions = {
        scrollHeight: body == null ? 0 : body.scrollHeight,
        offsetHeight: body == null ? 0 : body.offsetHeight,
    };
    var htmlDimensions = {
        clientHeight: html == null ? 0 : html.clientHeight,
        clientWidth: html == null ? 0 : html.clientWidth,
        scrollHeight: html == null ? 0 : html.scrollHeight,
        scrollWidth: html == null ? 0 : html.scrollWidth,
        offsetHeight: html == null ? 0 : html.offsetHeight,
    };
    var windowDimensions = {
        innerWidth: window.innerWidth,
        innerHeight: window.innerHeight,
        isLandscape: window.matchMedia('(orientation: landscape)').matches,
        outerHeight: window.outerHeight === 0 ? htmlDimensions.clientHeight : window.outerHeight,
        outerWidth: window.outerWidth === 0 ? htmlDimensions.clientWidth : window.outerWidth,
        devicePixelRatio: window.devicePixelRatio,
        screenWidth: window.screen.width,
        screenHeight: window.screen.height,
    };
    return {
        dimensions: {
            body: bodyDimensions,
            html: htmlDimensions,
            window: windowDimensions,
        },
    };
}
exports.default = getScreenDimensions;
