// webdriver-image-comparison/build/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var fs_extra_1 = require("fs-extra");
var options_1 = require("./helpers/options");
var constants_1 = require("./helpers/constants");
var BaseClass = (function () {
    function BaseClass(options) {
        this.defaultOptions = (0, options_1.defaultOptions)(options);
        var baselineFolder;
        var baseFolder;
        if (typeof options.baselineFolder === 'function') {
            baselineFolder = options.baselineFolder(options);
        }
        else {
            baselineFolder = (0, path_1.normalize)(options.baselineFolder || constants_1.FOLDERS.DEFAULT.BASE);
        }
        if (typeof options.screenshotPath === 'function') {
            baseFolder = options.screenshotPath(options);
        }
        else {
            baseFolder = (0, path_1.normalize)(options.screenshotPath || constants_1.FOLDERS.DEFAULT.SCREENSHOTS);
        }
        this.folders = {
            actualFolder: (0, path_1.join)(baseFolder, constants_1.FOLDERS.ACTUAL),
            baselineFolder: baselineFolder,
            diffFolder: (0, path_1.join)(baseFolder, constants_1.FOLDERS.DIFF),
        };
        if (options.clearRuntimeFolder) {
            console.log('\n\n\n##############################');
            console.log('!!CLEARING!!');
            console.log('##############################\n\n\n');
            (0, fs_extra_1.removeSync)(this.folders.actualFolder);
            (0, fs_extra_1.removeSync)(this.folders.diffFolder);
        }
    }
    return BaseClass;
}());
exports.default = BaseClass;
