// webdriver-image-comparison/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTabbablePage = exports.checkFullPageScreen = exports.checkElement = exports.checkScreen = exports.saveTabbablePage = exports.saveFullPageScreen = exports.saveElement = exports.saveScreen = exports.BaseClass = void 0;
var base_1 = require("./base");
exports.BaseClass = base_1.default;
var saveScreen_1 = require("./commands/saveScreen");
exports.saveScreen = saveScreen_1.default;
var saveElement_1 = require("./commands/saveElement");
exports.saveElement = saveElement_1.default;
var saveFullPageScreen_1 = require("./commands/saveFullPageScreen");
exports.saveFullPageScreen = saveFullPageScreen_1.default;
var saveTabbablePage_1 = require("./commands/saveTabbablePage");
exports.saveTabbablePage = saveTabbablePage_1.default;
var checkScreen_1 = require("./commands/checkScreen");
exports.checkScreen = checkScreen_1.default;
var checkElement_1 = require("./commands/checkElement");
exports.checkElement = checkElement_1.default;
var checkFullPageScreen_1 = require("./commands/checkFullPageScreen");
exports.checkFullPageScreen = checkFullPageScreen_1.default;
var checkTabbablePage_1 = require("./commands/checkTabbablePage");
exports.checkTabbablePage = checkTabbablePage_1.default;
