// webdriver-image-comparison/build/helpers/constants.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedIosBezelDevices = exports.IOS_OFFSETS = exports.ANDROID_OFFSETS = exports.DEFAULT_TABBABLE_OPTIONS = exports.FULL_PAGE_SCROLL_TIMEOUT = exports.DEFAULT_RESIZE_DIMENSIONS = exports.CUSTOM_CSS_ID = exports.DESKTOP = exports.DEFAULT_SHADOW = exports.FOLDERS = exports.PLATFORMS = exports.DEFAULT_FORMAT_STRING = void 0;
exports.DEFAULT_FORMAT_STRING = '{tag}-{browserName}-{width}x{height}-dpr-{dpr}';
exports.PLATFORMS = {
    ANDROID: 'android',
    IOS: 'ios',
};
exports.FOLDERS = {
    ACTUAL: 'actual',
    DIFF: 'diff',
    TEMP_FULL_SCREEN: 'tempFullScreen',
    DEFAULT: {
        BASE: './wic/baseline/',
        SCREENSHOTS: '.tmp/',
    },
};
exports.DEFAULT_SHADOW = {
    ADDRESS_BAR: 6,
    TOOL_BAR: 6,
};
exports.DESKTOP = 'desktop';
exports.CUSTOM_CSS_ID = 'pic-css';
exports.DEFAULT_RESIZE_DIMENSIONS = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
};
exports.FULL_PAGE_SCROLL_TIMEOUT = 1500;
exports.DEFAULT_TABBABLE_OPTIONS = {
    circle: {
        backgroundColor: '#ff0000',
        borderColor: '#000',
        borderWidth: 1,
        fontColor: '#fff',
        fontFamily: 'Arial',
        fontSize: 10,
        size: 10,
        showNumber: true,
    },
    line: {
        color: '#000',
        width: 1,
    },
};
var ANDROID_DEFAULT_OFFSETS = {
    STATUS_BAR: 24,
    ADDRESS_BAR: 56,
    TOOL_BAR: 48,
};
exports.ANDROID_OFFSETS = {
    6: __assign({}, ANDROID_DEFAULT_OFFSETS),
    7: __assign({}, ANDROID_DEFAULT_OFFSETS),
    8: __assign({}, ANDROID_DEFAULT_OFFSETS),
    9: __assign({}, ANDROID_DEFAULT_OFFSETS),
    10: __assign({}, ANDROID_DEFAULT_OFFSETS),
    11: __assign({}, ANDROID_DEFAULT_OFFSETS),
    12: __assign({}, ANDROID_DEFAULT_OFFSETS),
};
exports.IOS_OFFSETS = {
    IPHONE: {
        667: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        736: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        812: {
            LANDSCAPE: {
                SAFE_AREA: 44,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 300, y: 359, height: 9, width: 212 },
            },
            PORTRAIT: {
                SAFE_AREA: 44,
                STATUS_BAR: 44,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 118, y: 796, height: 9, width: 137 },
            },
        },
        844: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 312, y: 375, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 124, y: 829, height: 9, width: 143 },
            },
        },
        852: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 316, y: 378, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 125, y: 837, height: 9, width: 143 },
            },
        },
        896: {
            LANDSCAPE: {
                SAFE_AREA: 48,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 331, y: 399, height: 9, width: 234 },
            },
            PORTRAIT: {
                SAFE_AREA: 48,
                STATUS_BAR: 48,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 131, y: 881, height: 9, width: 152 },
            },
        },
        926: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 342, y: 413, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 136, y: 911, height: 9, width: 156 },
            },
        },
        932: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 345, y: 415, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 137, y: 917, height: 9, width: 158 },
            },
        },
    },
    IPAD: {
        1024: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        1080: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        1112: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        1133: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 408, y: 729, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 234, y: 1118, height: 9, width: 276 },
            },
        },
        1180: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 431, y: 805, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 272, y: 1165, height: 9, width: 276 },
            },
        },
        1194: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 438, y: 819, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 279, y: 1179, height: 9, width: 276 },
            },
        },
        1366: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 508, y: 1009, height: 9, width: 350 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 352, y: 1351, height: 9, width: 320 },
            },
        },
    },
};
exports.supportedIosBezelDevices = [
    'iphonex',
    'iphonexs',
    'iphonexsmax',
    'iphonexr',
    'iphone11',
    'iphone11pro',
    'iphone11promax',
    'iphone12',
    'iphone12mini',
    'iphone12pro',
    'iphone12promax',
    'iphone13',
    'iphone13mini',
    'iphone13pro',
    'iphone13promax',
    'iphone14',
    'iphone14plus',
    'iphone14pro',
    'iphone14promax',
    'ipadmini',
    'ipadair',
    'ipadpro11',
    'ipadpro129',
];
