// webdriver-image-comparison/build/commands/saveScreen.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var screenshots_1 = require("../methods/screenshots");
var images_1 = require("../methods/images");
var beforeScreenshot_1 = require("../helpers/beforeScreenshot");
var afterScreenshot_1 = require("../helpers/afterScreenshot");
var rectangles_1 = require("../methods/rectangles");
function saveScreen(methods, instanceData, folders, tag, saveScreenOptions) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, addressBarShadowPadding, addIOSBezelCorners, formatImageName, logLevel, savePerInstance, toolBarShadowPadding, disableCSSAnimation, hideScrollBars, hideElements, removeElements, beforeOptions, enrichedInstanceData, browserName, browserVersion, deviceName, _b, devicePixelRatio, innerHeight, innerWidth, isLandscape, outerHeight, outerWidth, screenHeight, screenWidth, isAndroidChromeDriverScreenshot, isAndroidNativeWebScreenshot, isIos, isMobile, isTestInBrowser, logName, name, platformName, platformVersion, base64Image, screenRectangleOptions, rectangles, croppedBase64Image, afterOptions;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = saveScreenOptions.wic, addressBarShadowPadding = _a.addressBarShadowPadding, addIOSBezelCorners = _a.addIOSBezelCorners, formatImageName = _a.formatImageName, logLevel = _a.logLevel, savePerInstance = _a.savePerInstance, toolBarShadowPadding = _a.toolBarShadowPadding;
                    disableCSSAnimation = 'disableCSSAnimation' in saveScreenOptions.method
                        ? saveScreenOptions.method.disableCSSAnimation
                        : saveScreenOptions.wic.disableCSSAnimation;
                    hideScrollBars = 'hideScrollBars' in saveScreenOptions.method ? saveScreenOptions.method.hideScrollBars : saveScreenOptions.wic.hideScrollBars;
                    hideElements = saveScreenOptions.method.hideElements || [];
                    removeElements = saveScreenOptions.method.removeElements || [];
                    beforeOptions = {
                        instanceData: instanceData,
                        addressBarShadowPadding: addressBarShadowPadding,
                        disableCSSAnimation: disableCSSAnimation,
                        hideElements: hideElements,
                        logLevel: logLevel,
                        noScrollBars: hideScrollBars,
                        removeElements: removeElements,
                        toolBarShadowPadding: toolBarShadowPadding,
                    };
                    return [4, (0, beforeScreenshot_1.default)(methods.executor, beforeOptions)];
                case 1:
                    enrichedInstanceData = _c.sent();
                    browserName = enrichedInstanceData.browserName, browserVersion = enrichedInstanceData.browserVersion, deviceName = enrichedInstanceData.deviceName, _b = enrichedInstanceData.dimensions.window, devicePixelRatio = _b.devicePixelRatio, innerHeight = _b.innerHeight, innerWidth = _b.innerWidth, isLandscape = _b.isLandscape, outerHeight = _b.outerHeight, outerWidth = _b.outerWidth, screenHeight = _b.screenHeight, screenWidth = _b.screenWidth, isAndroidChromeDriverScreenshot = enrichedInstanceData.isAndroidChromeDriverScreenshot, isAndroidNativeWebScreenshot = enrichedInstanceData.isAndroidNativeWebScreenshot, isIos = enrichedInstanceData.isIos, isMobile = enrichedInstanceData.isMobile, isTestInBrowser = enrichedInstanceData.isTestInBrowser, logName = enrichedInstanceData.logName, name = enrichedInstanceData.name, platformName = enrichedInstanceData.platformName, platformVersion = enrichedInstanceData.platformVersion;
                    return [4, (0, screenshots_1.takeBase64Screenshot)(methods.screenShot)];
                case 2:
                    base64Image = _c.sent();
                    screenRectangleOptions = {
                        devicePixelRatio: devicePixelRatio,
                        innerHeight: innerHeight,
                        innerWidth: innerWidth,
                        isAndroidChromeDriverScreenshot: isAndroidChromeDriverScreenshot,
                        isAndroidNativeWebScreenshot: isAndroidNativeWebScreenshot,
                        isIos: isIos,
                        isLandscape: isLandscape,
                    };
                    rectangles = (0, rectangles_1.determineScreenRectangles)(base64Image, screenRectangleOptions);
                    return [4, (0, images_1.makeCroppedBase64Image)({
                            addIOSBezelCorners: addIOSBezelCorners,
                            base64Image: base64Image,
                            deviceName: deviceName,
                            devicePixelRatio: devicePixelRatio,
                            isIos: isIos,
                            isLandscape: isLandscape,
                            logLevel: logLevel,
                            rectangles: rectangles,
                        })];
                case 3:
                    croppedBase64Image = _c.sent();
                    afterOptions = {
                        actualFolder: folders.actualFolder,
                        base64Image: croppedBase64Image,
                        disableCSSAnimation: disableCSSAnimation,
                        filePath: {
                            browserName: browserName,
                            deviceName: deviceName,
                            isMobile: isMobile,
                            savePerInstance: savePerInstance,
                        },
                        fileName: {
                            browserName: browserName,
                            browserVersion: browserVersion,
                            deviceName: deviceName,
                            devicePixelRatio: devicePixelRatio,
                            formatImageName: formatImageName,
                            isMobile: isMobile,
                            isTestInBrowser: isTestInBrowser,
                            logName: logName,
                            name: name,
                            outerHeight: outerHeight,
                            outerWidth: outerWidth,
                            platformName: platformName,
                            platformVersion: platformVersion,
                            screenHeight: screenHeight,
                            screenWidth: screenWidth,
                            tag: tag,
                        },
                        hideElements: hideElements,
                        hideScrollBars: hideScrollBars,
                        isLandscape: isLandscape,
                        logLevel: logLevel,
                        platformName: instanceData.platformName,
                        removeElements: removeElements,
                    };
                    return [2, (0, afterScreenshot_1.default)(methods.executor, afterOptions)];
            }
        });
    });
}
exports.default = saveScreen;
